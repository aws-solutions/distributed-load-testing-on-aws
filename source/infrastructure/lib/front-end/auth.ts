// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  UserPool,
  CfnUserPool,
  UserPoolClient,
  ClientAttributes,
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  CfnUserPoolUser,
} from "aws-cdk-lib/aws-cognito";
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from "aws-cdk-lib/aws-iam";
import { Aws, ArnFormat, CfnResource, Duration, RemovalPolicy, Stack } from "aws-cdk-lib";
import { CfnPolicy } from "aws-cdk-lib/aws-iot";
import { Construct } from "constructs";

/**
 * CognitoAuthConstruct props
 *
 * @interface CognitoAuthConstructProps
 */
export interface CognitoAuthConstructProps {
  adminEmail: string;
  adminName: string;
  apiId: string;
  webAppURL: string;
  scenariosBucketArn: string;
}

export class CognitoAuthConstruct extends Construct {
  cognitoIdentityPoolId: string;
  cognitoUserPoolClientId: string;
  cognitoUserPoolId: string;
  public iotPolicy: CfnPolicy;

  constructor(scope: Construct, id: string, props: CognitoAuthConstructProps) {
    super(scope, id);

    const dltIotPolicy = new CfnPolicy(this, "IoT-Policy", {
      policyDocument: new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Connect"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "client",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Subscribe"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "topicfilter",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Receive"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "topic",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
        ],
      }),
    });
    this.iotPolicy = dltIotPolicy;
    dltIotPolicy.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "Cannot specify the resource to attach policy to identity",
        },
      ],
    });

    const cognitoUserPool = new UserPool(this, "DLTUserPool", {
      autoVerify: {
        email: true,
      },
      passwordPolicy: {
        minLength: 12,
        requireLowercase: true,
        requireDigits: true,
        requireSymbols: true,
        requireUppercase: true,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true,
        username: true,
      },
      standardAttributes: {
        email: {
          required: true,
        },
      },
      userInvitation: {
        emailSubject: "Welcome to Distributed Load Testing",
        emailBody: `
                <p>
                   Please use the credentials below to login to the Distributed Load Testing console.
                </p>
                <p>
                    Username: <strong>{username}</strong>
                </p>
                <p>
                    Password: <strong>{####}</strong>
                </p>
                <p>
                    Console: <strong>${props.webAppURL}/</strong>
                </p>
              `,
        smsMessage: "Your username is {username} and temporary password is {####}.",
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
    });
    (cognitoUserPool.node.defaultChild as CfnUserPool).userPoolAddOns = { advancedSecurityMode: "ENFORCED" };
    this.cognitoUserPoolId = cognitoUserPool.userPoolId;

    const clientWriteAttributes = new ClientAttributes().withStandardAttributes({
      address: true,
      email: true,
      phoneNumber: true,
    });

    const cognitoUserPoolClient = new UserPoolClient(this, "DLTUserPoolClient", {
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`,
      userPool: cognitoUserPool,
      generateSecret: false,
      writeAttributes: clientWriteAttributes,
      refreshTokenValidity: Duration.days(1),
    });

    this.cognitoUserPoolClientId = cognitoUserPoolClient.userPoolClientId;

    const cognitoIdentityPool = new CfnIdentityPool(this, "DLTIdentityPool", {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: this.cognitoUserPoolClientId,
          providerName: cognitoUserPool.userPoolProviderName,
        },
      ],
    });

    this.cognitoIdentityPoolId = cognitoIdentityPool.ref;

    const apiProdExecuteArn = Stack.of(this).formatArn({
      service: "execute-api",
      resource: props.apiId,
      resourceName: "prod/*",
    });
    const cognitoAuthorizedRole = new Role(this, "DLTCognitoAuthorizedRole", {
      assumedBy: new FederatedPrincipal(
        "cognito-identity.amazonaws.com",
        {
          StringEquals: { "cognito-identity.amazonaws.com:aud": this.cognitoIdentityPoolId },
          "ForAnyValue:StringLike": { "cognito-identity.amazonaws.com:amr": "authenticated" },
        },
        "sts:AssumeRoleWithWebIdentity"
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        InvokeApiPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["execute-api:Invoke"],
              resources: [apiProdExecuteArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:PutObject", "s3:GetObject"],
              resources: [`${props.scenariosBucketArn}/public/*`, `${props.scenariosBucketArn}/cloudWatchImages/*`],
            }),
          ],
        }),
        IoTPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:AttachPrincipalPolicy"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Connect"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "client",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Subscribe"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "topicfilter",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Receive"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "topic",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
          ],
        }),
      },
    });
    (cognitoAuthorizedRole.node.defaultChild as CfnResource).addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "iot:AttachPrincipalPolicy does not allow for resource specification",
        },
      ],
    });

    const cognitoUnauthorizedRole = new Role(this, "DLTCognitoUnauthorizedRole", {
      assumedBy: new FederatedPrincipal(
        "cognito-identity.amazonaws.com",
        {
          StringEquals: { "cognito-identity.amazonaws.com:aud": this.cognitoIdentityPoolId },
          "ForAnyValue:StringLike": { "cognito-identity.amazonaws.com:amr": "unauthenticated" },
        },
        "sts:AssumeRoleWithWebIdentity"
      ),
    });

    new CfnIdentityPoolRoleAttachment(this, "CognitoAttachRole", {
      identityPoolId: this.cognitoIdentityPoolId,
      roles: {
        unauthenticated: cognitoUnauthorizedRole.roleArn,
        authenticated: cognitoAuthorizedRole.roleArn,
      },
    });

    new CfnUserPoolUser(this, "CognitoUser", {
      desiredDeliveryMediums: ["EMAIL"],
      forceAliasCreation: true,
      userAttributes: [
        { name: "email", value: props.adminEmail },
        { name: "nickname", value: props.adminName },
        { name: "email_verified", value: "true" },
      ],
      username: props.adminName,
      userPoolId: this.cognitoUserPoolId,
    });
  }
}
