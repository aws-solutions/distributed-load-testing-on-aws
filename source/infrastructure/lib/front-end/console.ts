// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CloudFrontToS3 } from "@aws-solutions-constructs/aws-cloudfront-s3";
import { Bucket, IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { BucketDeployment, Source as S3Source } from "aws-cdk-lib/aws-s3-deployment";
import * as path from "path";

export interface DLTConsoleConstructProps {
  readonly s3LogsBucket: Bucket;
  readonly solutionId: string;
}

export class DLTConsoleConstruct extends Construct {
  public webAppURL: string;
  public consoleBucketArn: string;
  public consoleBucket: IBucket;

  constructor(scope: Construct, id: string, props: DLTConsoleConstructProps) {
    super(scope, id);

    // Create the standard region CloudFront distribution
    const dltS3CloudFrontDist = new CloudFrontToS3(this, "DLTDashboardS3", {
      bucketProps: {
        serverAccessLogsBucket: props.s3LogsBucket,
        serverAccessLogsPrefix: "console-bucket-access/",
      },
      cloudFrontDistributionProps: {
        comment: "Website distribution for the Distributed Load Testing solution",
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: "/index.html" },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: "/index.html" },
        ],
        httpVersion: "http2",
        logBucket: props.s3LogsBucket,
        logFilePrefix: "cloudfront-logs/",
      },
      insertHttpSecurityHeaders: false,
    });

    this.consoleBucket = dltS3CloudFrontDist.s3BucketInterface;
    this.webAppURL = `https://${dltS3CloudFrontDist.cloudFrontWebDistribution.domainName}`;
    this.consoleBucketArn = dltS3CloudFrontDist.s3BucketInterface.bucketArn;

    new BucketDeployment(this, "DeployWebsite", {
      sources: [S3Source.asset(path.join(__dirname, "../../../console/build"))], // build react app before cdk deploy
      destinationBucket: this.consoleBucket,
      exclude: ["assets/aws_config.js"],
    });
  }
}
