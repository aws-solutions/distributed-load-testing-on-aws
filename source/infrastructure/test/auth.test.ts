// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";

import { CognitoAuthConstruct } from "../lib/front-end/auth";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });

  const auth = new CognitoAuthConstruct(stack, "TestAuth", {
    adminEmail: "test@test.com",
    adminName: "testname",
    apiId: "apiId12345",
    cloudFrontDomainName: "test.com",
    scenariosBucketArn: "arn:aws:s3:::DOC-EXAMPLE-BUCKET",
    existingCognitoPoolId: "",
  });

  const auth2 = new CognitoAuthConstruct(stack, "TestAuth2", {
    adminEmail: "test@test.com",
    adminName: "testname",
    apiId: "apiId12345",
    cloudFrontDomainName: "test.com",
    scenariosBucketArn: "arn:aws:s3:::DOC-EXAMPLE-BUCKET",
    existingCognitoPoolId: "us-east-1_123456789",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
  expect(auth.cognitoIdentityPoolId).toBeDefined();
  expect(auth.cognitoUserPoolClientId).toBeDefined();
  expect(auth.cognitoUserPoolId).toBeDefined();

  expect(auth2.cognitoIdentityPoolId).toBeDefined();
  expect(auth2.cognitoUserPoolClientId).toBeDefined();
  expect(auth2.cognitoUserPoolId).toBeDefined();
});
