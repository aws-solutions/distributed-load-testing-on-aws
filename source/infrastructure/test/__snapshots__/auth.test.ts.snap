// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Resources": {
    "TestAuthCognitoAttachRole3C8C2843": {
      "Properties": {
        "IdentityPoolId": {
          "Ref": "TestAuthDLTIdentityPoolCE53676D",
        },
        "Roles": {
          "authenticated": {
            "Fn::GetAtt": [
              "TestAuthDLTCognitoAuthorizedRole00B36330",
              "Arn",
            ],
          },
          "unauthenticated": {
            "Fn::GetAtt": [
              "TestAuthDLTCognitoUnauthorizedRole3D399466",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "TestAuthCognitoUser7FC8B6DF": {
      "Properties": {
        "DesiredDeliveryMediums": [
          "EMAIL",
        ],
        "ForceAliasCreation": true,
        "UserAttributes": [
          {
            "Name": "email",
            "Value": "test@test.com",
          },
          {
            "Name": "nickname",
            "Value": "testname",
          },
          {
            "Name": "email_verified",
            "Value": "true",
          },
        ],
        "UserPoolId": {
          "Ref": "TestAuthDLTUserPool047D7286",
        },
        "Username": "testname",
      },
      "Type": "AWS::Cognito::UserPoolUser",
    },
    "TestAuthDLTCognitoAuthorizedRole00B36330": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "iot:AttachPrincipalPolicy does not allow for resource specification",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "TestAuthDLTIdentityPoolCE53676D",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              " Identity Pool authenticated role",
            ],
          ],
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "execute-api:Invoke",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":execute-api:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":apiId12345/prod/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:GetObject",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:s3:::DOC-EXAMPLE-BUCKET/public/*",
                    "arn:aws:s3:::DOC-EXAMPLE-BUCKET/cloudWatchImages/*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "InvokeApiPolicy",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "iot:AttachPrincipalPolicy",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "iot:Connect",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":client/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": "iot:Subscribe",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":topicfilter/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": "iot:Receive",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":topic/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestAuthDLTCognitoUnauthorizedRole3D399466": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "unauthenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "TestAuthDLTIdentityPoolCE53676D",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestAuthDLTIdentityPoolCE53676D": {
      "Properties": {
        "AllowUnauthenticatedIdentities": false,
        "CognitoIdentityProviders": [
          {
            "ClientId": {
              "Ref": "TestAuthDLTUserPoolClient68D1C3B2",
            },
            "ProviderName": {
              "Fn::GetAtt": [
                "TestAuthDLTUserPool047D7286",
                "ProviderName",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Cognito::IdentityPool",
    },
    "TestAuthDLTUserPool047D7286": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": true,
          "InviteMessageTemplate": {
            "EmailMessage": "
                <p>
                   Please use the credentials below to login to the Distributed Load Testing console.
                </p>
                <p>
                    Username: <strong>{username}</strong>
                </p>
                <p>
                    Password: <strong>{####}</strong>
                </p>
                <p>
                    Console: <strong>https://test.com/</strong>
                </p>
              ",
            "EmailSubject": "Welcome to Distributed Load Testing",
            "SMSMessage": "Your username is {username} and temporary password is {####}.",
          },
        },
        "AliasAttributes": [
          "email",
        ],
        "AutoVerifiedAttributes": [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": {
          "PasswordPolicy": {
            "MinimumLength": 12,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true,
          },
        },
        "Schema": [
          {
            "Mutable": true,
            "Name": "email",
            "Required": true,
          },
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolAddOns": {
          "AdvancedSecurityMode": "ENFORCED",
        },
        "UserPoolName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-user-pool",
            ],
          ],
        },
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "TestAuthDLTUserPoolClient68D1C3B2": {
      "Properties": {
        "AllowedOAuthFlows": [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": [
          "https://example.com",
        ],
        "ClientName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-userpool-client",
            ],
          ],
        },
        "GenerateSecret": false,
        "RefreshTokenValidity": 1440,
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "TokenValidityUnits": {
          "RefreshToken": "minutes",
        },
        "UserPoolId": {
          "Ref": "TestAuthDLTUserPool047D7286",
        },
        "WriteAttributes": [
          "address",
          "email",
          "phone_number",
        ],
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "TestAuthIoTPolicyF897DD43": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "Cannot specify the resource to attach policy to identity",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:Connect",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":client/*",
                  ],
                ],
              },
            },
            {
              "Action": "iot:Subscribe",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":topicfilter/*",
                  ],
                ],
              },
            },
            {
              "Action": "iot:Receive",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IoT::Policy",
    },
  },
}
`;
